ICE: `can't get the span of an arbitrary parsed attribute`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at /rustc-dev/3b97f1308ff72016a4aaa93fbe6d09d4d6427815/compiler/rustc_hir/src/hir.rs:1305:18: 'can't get the span of an arbitrary parsed attribute: Parsed(Inline(Hint, /tmp/im/a.rs:1:1: 1:10 (#0)))'', 'thread 'rustc' panicked at /rustc-dev/3b97f1308ff72016a4aaa93fbe6d09d4d6427815/compiler/rustc_hir/src/hir.rs:1305:18: 'can't get the span of an arbitrary parsed attribute: Parsed(Inline(Hint, /tmp/im/a.rs:1:1: 1:10 (#0)))''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
#[inline]
#[track_caller]
macro_rules! contained {
    () => {};
}

````

original:
````rust
#[inline]
#[track_caller]
macro_rules! contained { () => {} }

pub fn main() {}

````

Version information
````
rustc 1.89.0-nightly (3b97f1308 2025-06-20)
binary: rustc
commit-hash: 3b97f1308ff72016a4aaa93fbe6d09d4d6427815
commit-date: 2025-06-20
host: x86_64-unknown-linux-gnu
release: 1.89.0-nightly
LLVM version: 20.1.7

````



Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<details><summary><strong>Program output</strong></summary>
<p>

```
warning: unused macro definition: `contained`
 --> /tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs:3:14
  |
3 | macro_rules! contained {
  |              ^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs`

error[E0518]: attribute should be applied to function or closure
 --> /tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs:1:1
  |
1 | #[inline]
  | ^^^^^^^^^ not a function or closure

warning: `#[track_caller]` is ignored on struct fields, match arms and macro defs
 --> /tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs:2:1
  |
2 | #[track_caller]
  | ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default


thread 'rustc' panicked at /rustc-dev/3b97f1308ff72016a4aaa93fbe6d09d4d6427815/compiler/rustc_hir/src/hir.rs:1305:18:
can't get the span of an arbitrary parsed attribute: Parsed(Inline(Hint, /tmp/icemaker_global_tempdir.h2zBL4aoaUWH/rustc_testrunner_tmpdir_reporting.EsRoILp2XH1B/mvce.rs:1:1: 1:10 (#0)))
stack backtrace:
   0:     0x796ff7ce3033 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hdb1beb1efd036030
   1:     0x796ff8402d3f - core::fmt::write::h6fde5d988b046e43
   2:     0x796ff7cd8cc3 - std::io::Write::write_fmt::hfdc56ad9972213fb
   3:     0x796ff7ce2e92 - std::sys::backtrace::BacktraceLock::print::h7d50fbfbc01ef440
   4:     0x796ff7ce698a - std::panicking::default_hook::{{closure}}::h2d0b345b2d72c7b1
   5:     0x796ff7ce650f - std::panicking::default_hook::he0e1084842380634
   6:     0x796ff6ddc397 - std[6b85dd0cacdad3fb]::panicking::update_hook::<alloc[e197c52ffb0c953d]::boxed::Box<rustc_driver_impl[b5010b6a00a10337]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x796ff7ce71db - std::panicking::rust_panic_with_hook::h551040e3eccc63b8
   8:     0x796ff7ce6eda - std::panicking::begin_panic_handler::{{closure}}::h82732907cb122ecc
   9:     0x796ff7ce3509 - std::sys::backtrace::__rust_end_short_backtrace::h8d0b6d7b81e6f1ec
  10:     0x796ff7ce6bbd - __rustc[e49b1424bdfe04fc]::rust_begin_unwind
  11:     0x796ff448fbc0 - core::panicking::panic_fmt::h2de75984da6179f3
  12:     0x796ff8ae0d9f - <rustc_passes[ed7886fbdff4a823]::check_attr::CheckAttrVisitor>::check_attributes
  13:     0x796ff8ad7c86 - <rustc_passes[ed7886fbdff4a823]::check_attr::CheckAttrVisitor as rustc_hir[cd02d9627a3ada87]::intravisit::Visitor>::visit_item
  14:     0x796ff8ad71bf - rustc_passes[ed7886fbdff4a823]::check_attr::check_mod_attrs
  15:     0x796ff8ad7061 - rustc_query_impl[7d775a893a945633]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7d775a893a945633]::query_impl::check_mod_attrs::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cffb9c37e6e9a24b]::query::erase::Erased<[u8; 0usize]>>
  16:     0x796ff928d058 - rustc_query_system[2228cb53b85e8de8]::query::plumbing::try_execute_query::<rustc_query_impl[7d775a893a945633]::DynamicConfig<rustc_query_system[2228cb53b85e8de8]::query::caches::DefaultCache<rustc_span[fb51e5d36bb23ffe]::def_id::LocalModDefId, rustc_middle[cffb9c37e6e9a24b]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[7d775a893a945633]::plumbing::QueryCtxt, false>
  17:     0x796ff928cd7d - rustc_query_impl[7d775a893a945633]::query_impl::check_mod_attrs::get_query_non_incr::__rust_end_short_backtrace
  18:     0x796ff928ca1b - <rustc_middle[cffb9c37e6e9a24b]::ty::context::TyCtxt>::par_hir_for_each_module::<rustc_interface[400f48cd83754571]::passes::run_required_analyses::{closure#0}::{closure#0}::{closure#1}::{closure#0}>::{closure#0}
  19:     0x796ff8ed7e8b - rustc_interface[400f48cd83754571]::passes::analysis
  20:     0x796ff8ed7bad - rustc_query_impl[7d775a893a945633]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7d775a893a945633]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[cffb9c37e6e9a24b]::query::erase::Erased<[u8; 0usize]>>
  21:     0x796ff94a1d3c - rustc_query_system[2228cb53b85e8de8]::query::plumbing::try_execute_query::<rustc_query_impl[7d775a893a945633]::DynamicConfig<rustc_query_system[2228cb53b85e8de8]::query::caches::SingleCache<rustc_middle[cffb9c37e6e9a24b]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[7d775a893a945633]::plumbing::QueryCtxt, false>
  22:     0x796ff94a1a12 - rustc_query_impl[7d775a893a945633]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  23:     0x796ff960ee96 - rustc_interface[400f48cd83754571]::passes::create_and_enter_global_ctxt::<core[913fe49cdece82e8]::option::Option<rustc_interface[400f48cd83754571]::queries::Linker>, rustc_driver_impl[b5010b6a00a10337]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  24:     0x796ff9627963 - rustc_interface[400f48cd83754571]::interface::run_compiler::<(), rustc_driver_impl[b5010b6a00a10337]::run_compiler::{closure#0}>::{closure#1}
  25:     0x796ff9651402 - std[6b85dd0cacdad3fb]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[400f48cd83754571]::util::run_in_thread_with_globals<rustc_interface[400f48cd83754571]::util::run_in_thread_pool_with_globals<rustc_interface[400f48cd83754571]::interface::run_compiler<(), rustc_driver_impl[b5010b6a00a10337]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  26:     0x796ff96510e6 - <<std[6b85dd0cacdad3fb]::thread::Builder>::spawn_unchecked_<rustc_interface[400f48cd83754571]::util::run_in_thread_with_globals<rustc_interface[400f48cd83754571]::util::run_in_thread_pool_with_globals<rustc_interface[400f48cd83754571]::interface::run_compiler<(), rustc_driver_impl[b5010b6a00a10337]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[913fe49cdece82e8]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  27:     0x796ff964f73d - std::sys::pal::unix::thread::Thread::new::thread_start::ha32bf3e026b53de8
  28:     0x796ff30a57eb - <unknown>
  29:     0x796ff312918c - <unknown>
  30:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.89.0-nightly (3b97f1308 2025-06-20) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [check_mod_attrs] checking attributes in top-level module
#1 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors; 2 warnings emitted

Some errors have detailed explanations: E0518, E0601.
For more information about an error, try `rustc --explain E0518`.

```

</p>
</details>

<!--
query stack:
#0 [check_mod_attrs] checking attributes in top-level module
#1 [analysis] running analysis passes on this crate

-->

on current stable:
```
warning: unused macro definition: `contained`
 --> a.rs:3:14
  |
3 | macro_rules! contained { () => {} }
  |              ^^^^^^^^^
  |
  = note: `#[warn(unused_macros)]` on by default

error[E0518]: attribute should be applied to function or closure
 --> a.rs:1:1
  |
1 | #[inline]
  | ^^^^^^^^^ not a function or closure

warning: `#[inline]` is ignored on struct fields, match arms and macro defs
 --> a.rs:1:1
  |
1 | #[inline]
  | ^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default

warning: `#[track_caller]` is ignored on struct fields, match arms and macro defs
 --> a.rs:1:1
  |
1 | #[inline]
  | ^^^^^^^^^

warning: `#[track_caller]` is ignored on struct fields, match arms and macro defs
 --> a.rs:2:1
  |
2 | #[track_caller]
  | ^^^^^^^^^^^^^^^
```

--- a/compiler/rustc_ast_lowering/Cargo.toml
+++ b/compiler/rustc_ast_lowering/Cargo.toml
@@ -11,6 +11,7 @@ doctest = false
 rustc_abi = { path = "../rustc_abi" }
 rustc_ast = { path = "../rustc_ast" }
 rustc_ast_pretty = { path = "../rustc_ast_pretty" }
+rustc_attr_data_structures = { path = "../rustc_attr_data_structures" }
 rustc_attr_parsing = { path = "../rustc_attr_parsing" }
 rustc_data_structures = { path = "../rustc_data_structures" }
 rustc_errors = { path = "../rustc_errors" }

--- a/compiler/rustc_ast_lowering/src/expr.rs
+++ b/compiler/rustc_ast_lowering/src/expr.rs
@@ -4,6 +4,7 @@ use std::sync::Arc;
 use rustc_ast::ptr::P as AstP;
 use rustc_ast::*;
 use rustc_ast_pretty::pprust::expr_to_string;
+use rustc_attr_data_structures::{AttributeKind, find_attr};
 use rustc_data_structures::stack::ensure_sufficient_stack;
 use rustc_hir as hir;
 use rustc_hir::HirId;
@@ -831,7 +832,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
     ) {
         if self.tcx.features().async_fn_track_caller()
             && let Some(attrs) = self.attrs.get(&outer_hir_id.local_id)
-            && attrs.into_iter().any(|attr| attr.has_name(sym::track_caller))
+            && find_attr!(*attrs, AttributeKind::TrackCaller(_))
         {
             let unstable_span = self.mark_span_with_reason(
                 DesugaringKind::Async,

--- a/compiler/rustc_attr_data_structures/src/attributes.rs
+++ b/compiler/rustc_attr_data_structures/src/attributes.rs
@@ -268,5 +268,8 @@ pub enum AttributeKind {
         /// Span of the attribute.
         span: Span,
     },
+
+    /// Represents `#[track_caller]`
+    TrackCaller(Span),
     // tidy-alphabetical-end
 }

--- a/compiler/rustc_attr_parsing/src/attributes/codegen_attrs.rs
+++ b/compiler/rustc_attr_parsing/src/attributes/codegen_attrs.rs
@@ -1,7 +1,7 @@
 use rustc_attr_data_structures::{AttributeKind, OptimizeAttr};
 use rustc_feature::{AttributeTemplate, template};
 use rustc_session::parse::feature_err;
-use rustc_span::{Span, sym};
+use rustc_span::{Span, Symbol, sym};
 
 use super::{AcceptMapping, AttributeOrder, AttributeParser, OnDuplicate, SingleAttributeParser};
 use crate::context::{AcceptContext, FinalizeContext, Stage};
@@ -11,7 +11,7 @@ use crate::session_diagnostics::NakedFunctionIncompatibleAttribute;
 pub(crate) struct OptimizeParser;
 
 impl<S: Stage> SingleAttributeParser<S> for OptimizeParser {
-    const PATH: &[rustc_span::Symbol] = &[sym::optimize];
+    const PATH: &[Symbol] = &[sym::optimize];
     const ATTRIBUTE_ORDER: AttributeOrder = AttributeOrder::KeepLast;
     const ON_DUPLICATE: OnDuplicate<S> = OnDuplicate::WarnButFutureError;
     const TEMPLATE: AttributeTemplate = template!(List: "size|speed|none");
@@ -44,7 +44,7 @@ impl<S: Stage> SingleAttributeParser<S> for OptimizeParser {
 pub(crate) struct ColdParser;
 
 impl<S: Stage> SingleAttributeParser<S> for ColdParser {
-    const PATH: &[rustc_span::Symbol] = &[sym::cold];
+    const PATH: &[Symbol] = &[sym::cold];
     const ATTRIBUTE_ORDER: AttributeOrder = AttributeOrder::KeepLast;
     const ON_DUPLICATE: OnDuplicate<S> = OnDuplicate::Warn;
     const TEMPLATE: AttributeTemplate = template!(Word);
@@ -163,6 +163,24 @@ impl<S: Stage> AttributeParser<S> for NakedParser {
         }
 
         Some(AttributeKind::Naked(span))
+    }
+}
+
+pub(crate) struct TrackCallerParser;
+
+impl<S: Stage> SingleAttributeParser<S> for TrackCallerParser {
+    const PATH: &[Symbol] = &[sym::track_caller];
+    const ATTRIBUTE_ORDER: AttributeOrder = AttributeOrder::KeepLast;
+    const ON_DUPLICATE: OnDuplicate<S> = OnDuplicate::Warn;
+    const TEMPLATE: AttributeTemplate = template!(Word);
+
+    fn convert(cx: &mut AcceptContext<'_, '_, S>, args: &ArgParser<'_>) -> Option<AttributeKind> {
+        if let Err(span) = args.no_args() {
+            cx.expected_no_args(span);
+            return None;
+        }
+
+        Some(AttributeKind::TrackCaller(cx.attr_span))
     }
 }
 

--- a/compiler/rustc_attr_parsing/src/context.rs
+++ b/compiler/rustc_attr_parsing/src/context.rs
@@ -15,7 +15,9 @@ use rustc_session::Session;
 use rustc_span::{DUMMY_SP, ErrorGuaranteed, Span, Symbol, sym};
 
 use crate::attributes::allow_unstable::{AllowConstFnUnstableParser, AllowInternalUnstableParser};
-use crate::attributes::codegen_attrs::{ColdParser, NakedParser, NoMangleParser, OptimizeParser};
+use crate::attributes::codegen_attrs::{
+    ColdParser, NakedParser, NoMangleParser, OptimizeParser, TrackCallerParser,
+};
 use crate::attributes::confusables::ConfusablesParser;
 use crate::attributes::deprecation::DeprecationParser;
 use crate::attributes::inline::{InlineParser, RustcForceInlineParser};
@@ -121,6 +123,7 @@ attribute_parsers!(
         Single<PubTransparentParser>,
         Single<RustcForceInlineParser>,
         Single<SkipDuringMethodDispatchParser>,
+        Single<TrackCallerParser>,
         Single<TransparencyParser>,
         // tidy-alphabetical-end
     ];

--- a/compiler/rustc_codegen_ssa/src/codegen_attrs.rs
+++ b/compiler/rustc_codegen_ssa/src/codegen_attrs.rs
@@ -95,17 +95,15 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
         // In these cases, we bail from performing further checks that are only meaningful for
         // functions (such as calling `fn_sig`, which ICEs if given a non-function). We also
         // report a delayed bug, just in case `check_attr` isn't doing its job.
-        let fn_sig = || {
+        let fn_sig = |attr_span| {
             use DefKind::*;
 
             let def_kind = tcx.def_kind(did);
             if let Fn | AssocFn | Variant | Ctor(..) = def_kind {
                 Some(tcx.fn_sig(did))
             } else {
-                tcx.dcx().span_delayed_bug(
-                    attr.span(),
-                    "this attribute can only be applied to functions",
-                );
+                tcx.dcx()
+                    .span_delayed_bug(attr_span, "this attribute can only be applied to functions");
                 None
             }
         };
@@ -141,6 +139,29 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
                             ),
                         });
                     }
+                }
+                AttributeKind::TrackCaller(attr_span) => {
+                    let is_closure = tcx.is_closure_like(did.to_def_id());
+
+                    if !is_closure
+                        && let Some(fn_sig) = fn_sig(*attr_span)
+                        && fn_sig.skip_binder().abi() != ExternAbi::Rust
+                    {
+                        tcx.dcx().emit_err(errors::RequiresRustAbi { span: *attr_span });
+                    }
+                    if is_closure
+                        && !tcx.features().closure_track_caller()
+                        && !attr_span.allows_unstable(sym::closure_track_caller)
+                    {
+                        feature_err(
+                            &tcx.sess,
+                            sym::closure_track_caller,
+                            *attr_span,
+                            "`#[track_caller]` on closures is currently unstable",
+                        )
+                        .emit();
+                    }
+                    codegen_fn_attrs.flags |= CodegenFnAttrFlags::TRACK_CALLER
                 }
                 _ => {}
             }
@@ -202,29 +223,6 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
                 }
             }
             sym::thread_local => codegen_fn_attrs.flags |= CodegenFnAttrFlags::THREAD_LOCAL,
-            sym::track_caller => {
-                let is_closure = tcx.is_closure_like(did.to_def_id());
-
-                if !is_closure
-                    && let Some(fn_sig) = fn_sig()
-                    && fn_sig.skip_binder().abi() != ExternAbi::Rust
-                {
-                    tcx.dcx().emit_err(errors::RequiresRustAbi { span: attr.span() });
-                }
-                if is_closure
-                    && !tcx.features().closure_track_caller()
-                    && !attr.span().allows_unstable(sym::closure_track_caller)
-                {
-                    feature_err(
-                        &tcx.sess,
-                        sym::closure_track_caller,
-                        attr.span(),
-                        "`#[track_caller]` on closures is currently unstable",
-                    )
-                    .emit();
-                }
-                codegen_fn_attrs.flags |= CodegenFnAttrFlags::TRACK_CALLER
-            }
             sym::export_name => {
                 if let Some(s) = attr.value_str() {
                     if s.as_str().contains('\0') {

--- a/compiler/rustc_hir_analysis/src/check/entry.rs
+++ b/compiler/rustc_hir_analysis/src/check/entry.rs
@@ -1,14 +1,15 @@
 use std::ops::Not;
 
 use rustc_abi::ExternAbi;
+use rustc_attr_data_structures::{AttributeKind, find_attr};
 use rustc_hir as hir;
 use rustc_hir::Node;
 use rustc_infer::infer::TyCtxtInferExt;
 use rustc_middle::span_bug;
 use rustc_middle::ty::{self, TyCtxt, TypingMode};
 use rustc_session::config::EntryFnType;
+use rustc_span::Span;
 use rustc_span::def_id::{CRATE_DEF_ID, DefId, LocalDefId};
-use rustc_span::{Span, sym};
 use rustc_trait_selection::error_reporting::InferCtxtErrorExt;
 use rustc_trait_selection::traits::{self, ObligationCause, ObligationCauseCode};
 
@@ -98,8 +99,10 @@ fn check_main_fn_ty(tcx: TyCtxt<'_>, main_def_id: DefId) {
         error = true;
     }
 
-    for attr in tcx.get_attrs(main_def_id, sym::track_caller) {
-        tcx.dcx().emit_err(errors::TrackCallerOnMain { span: attr.span(), annotated: main_span });
+    if let Some(attr_span) =
+        find_attr!(tcx.get_all_attrs(main_def_id), AttributeKind::TrackCaller(span) => *span)
+    {
+        tcx.dcx().emit_err(errors::TrackCallerOnMain { span: attr_span, annotated: main_span });
         error = true;
     }
 

--- a/compiler/rustc_lint/src/builtin.rs
+++ b/compiler/rustc_lint/src/builtin.rs
@@ -1184,11 +1184,11 @@ impl<'tcx> LateLintPass<'tcx> for UngatedAsyncFnTrackCaller {
         if fn_kind.asyncness().is_async()
             && !cx.tcx.features().async_fn_track_caller()
             // Now, check if the function has the `#[track_caller]` attribute
-            && let Some(attr) = cx.tcx.get_attr(def_id, sym::track_caller)
+            && let Some(attr_span) = find_attr!(cx.tcx.get_all_attrs(def_id), AttributeKind::TrackCaller(span) => *span)
         {
             cx.emit_span_lint(
                 UNGATED_ASYNC_FN_TRACK_CALLER,
-                attr.span(),
+                attr_span,
                 BuiltinUngatedAsyncFnTrackCaller { label: span, session: &cx.tcx.sess },
             );
         }

--- a/compiler/rustc_parse/src/validate_attr.rs
+++ b/compiler/rustc_parse/src/validate_attr.rs
@@ -301,6 +301,7 @@ fn emit_malformed_attribute(
             | sym::naked
             | sym::no_mangle
             | sym::must_use
+            | sym::track_caller
     ) {
         return;
     }

--- a/compiler/rustc_passes/src/check_attr.rs
+++ b/compiler/rustc_passes/src/check_attr.rs
@@ -722,9 +722,7 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
             // erroneously allowed it and some crates used it accidentally, to be compatible
             // with crates depending on them, we can't throw an error here.
             Target::Field | Target::Arm | Target::MacroDef => {
-                for attr in attrs {
-                    self.inline_attr_str_error_with_macro_def(hir_id, attr.span(), "track_caller");
-                }
+                self.inline_attr_str_error_with_macro_def(hir_id, attr_span, "track_caller");
             }
             _ => {
                 self.dcx().emit_err(errors::TrackedCallerWrongLocation {

--- a/compiler/rustc_passes/src/check_attr.rs
+++ b/compiler/rustc_passes/src/check_attr.rs
@@ -169,6 +169,9 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
                 Attribute::Parsed(AttributeKind::Naked(attr_span)) => {
                     self.check_naked(hir_id, *attr_span, span, target)
                 }
+                Attribute::Parsed(AttributeKind::TrackCaller(attr_span)) => {
+                    self.check_track_caller(hir_id, *attr_span, attrs, span, target)
+                }
                 Attribute::Parsed(
                     AttributeKind::BodyStability { .. }
                     | AttributeKind::ConstStabilityIndirect
@@ -205,9 +208,6 @@ impl<'tcx> CheckAttrVisitor<'tcx> {
                             self.check_target_feature(hir_id, attr, span, target, attrs)
                         }
                         [sym::thread_local, ..] => self.check_thread_local(attr, span, target),
-                        [sym::track_caller, ..] => {
-                            self.check_track_caller(hir_id, attr.span(), attrs, span, target)
-                        }
                         [sym::doc, ..] => self.check_doc_attrs(
                             attr,
                             attr_item.style,

--- a/src/tools/clippy/clippy_lints/src/eta_reduction.rs
+++ b/src/tools/clippy/clippy_lints/src/eta_reduction.rs
@@ -7,6 +7,7 @@ use clippy_utils::{
     get_path_from_caller_to_method_type, is_adjusted, is_no_std_crate, path_to_local, path_to_local_id,
 };
 use rustc_abi::ExternAbi;
+use rustc_attr_data_structures::{AttributeKind, find_attr};
 use rustc_errors::Applicability;
 use rustc_hir::{BindingMode, Expr, ExprKind, FnRetTy, GenericArgs, Param, PatKind, QPath, Safety, TyKind};
 use rustc_infer::infer::TyCtxtInferExt;
@@ -155,7 +156,7 @@ fn check_closure<'tcx>(cx: &LateContext<'tcx>, outer_receiver: Option<&Expr<'tcx
             let sig = match callee_ty_adjusted.kind() {
                 ty::FnDef(def, _) => {
                     // Rewriting `x(|| f())` to `x(f)` where f is marked `#[track_caller]` moves the `Location`
-                    if cx.tcx.has_attr(*def, sym::track_caller) {
+                    if find_attr!(cx.tcx.get_all_attrs(*def), AttributeKind::TrackCaller(..)) {
                         return;
                     }
 
@@ -236,7 +237,7 @@ fn check_closure<'tcx>(cx: &LateContext<'tcx>, outer_receiver: Option<&Expr<'tcx
         },
         ExprKind::MethodCall(path, self_, args, _) if check_inputs(typeck, body.params, Some(self_), args) => {
             if let Some(method_def_id) = typeck.type_dependent_def_id(body.value.hir_id)
-                && !cx.tcx.has_attr(method_def_id, sym::track_caller)
+                && !find_attr!(cx.tcx.get_all_attrs(method_def_id), AttributeKind::TrackCaller(..))
                 && check_sig(closure_sig, cx.tcx.fn_sig(method_def_id).skip_binder().skip_binder())
             {
                 let mut app = Applicability::MachineApplicable;

--- a/tests/ui/lint/unused/unused-attr-duplicate.stderr
+++ b/tests/ui/lint/unused/unused-attr-duplicate.stderr
@@ -90,18 +90,6 @@ LL | #[automatically_derived]
    | ^^^^^^^^^^^^^^^^^^^^^^^^
 
 error: unused attribute
-  --> $DIR/unused-attr-duplicate.rs:79:1
-   |
-LL | #[track_caller]
-   | ^^^^^^^^^^^^^^^ help: remove this attribute
-   |
-note: attribute also specified here
-  --> $DIR/unused-attr-duplicate.rs:78:1
-   |
-LL | #[track_caller]
-   | ^^^^^^^^^^^^^^^
-
-error: unused attribute
   --> $DIR/unused-attr-duplicate.rs:92:1
    |
 LL | #[export_name = "exported_symbol_name"]
@@ -276,6 +264,18 @@ note: attribute also specified here
    |
 LL | #[cold]
    | ^^^^^^^
+
+error: unused attribute
+  --> $DIR/unused-attr-duplicate.rs:79:1
+   |
+LL | #[track_caller]
+   | ^^^^^^^^^^^^^^^ help: remove this attribute
+   |
+note: attribute also specified here
+  --> $DIR/unused-attr-duplicate.rs:78:1
+   |
+LL | #[track_caller]
+   | ^^^^^^^^^^^^^^^
 
 error: unused attribute
   --> $DIR/unused-attr-duplicate.rs:98:1

--- a/tests/ui/rfcs/rfc-2091-track-caller/error-odd-syntax.stderr
+++ b/tests/ui/rfcs/rfc-2091-track-caller/error-odd-syntax.stderr
@@ -1,8 +1,12 @@
-error: malformed `track_caller` attribute input
+error[E0565]: malformed `track_caller` attribute input
   --> $DIR/error-odd-syntax.rs:1:1
    |
 LL | #[track_caller(1)]
-   | ^^^^^^^^^^^^^^^^^^ help: must be of the form: `#[track_caller]`
+   | ^^^^^^^^^^^^^^---^
+   | |             |
+   | |             didn't expect any arguments here
+   | help: must be of the form: `#[track_caller]`
 
 error: aborting due to 1 previous error
 
+For more information about this error, try `rustc --explain E0565`.

